File: app/src/main/java/com/pietrantuono/podcasts/addpodcast/presenter/AddPodcastPresenter.java
package com.pietrantuono.podcasts.addpodcast.presenter;

import android.support.annotation.Nullable;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.pietrantuono.podcasts.addpodcast.customviews.PodcastsAdapter;
import com.pietrantuono.podcasts.addpodcast.model.AddPodcastsModel;
import com.pietrantuono.podcasts.addpodcast.model.SearchResult;
import com.pietrantuono.podcasts.addpodcast.model.pojos.Podcast;
import com.pietrantuono.podcasts.addpodcast.view.AddPodcastFragmentMemento;
import com.pietrantuono.podcasts.addpodcast.view.AddPodcastView;
import com.pietrantuono.podcasts.GenericPresenter;
import com.pietrantuono.podcasts.addpodcast.view.ApiLevelChecker;

import rx.Observer;

public class AddPodcastPresenter implements GenericPresenter, PodcastsAdapter.OnSunscribeClickedListener, PodcastsAdapter.OnItemClickedClickedListener {
    public static final String TAG = (AddPodcastPresenter.class).getSimpleName();
    private final AddPodcastsModel addPodcastsModel;
    private final Observer<SearchResult> observer;
    private final ApiLevelChecker apiLevelChecker;
    @Nullable private AddPodcastView addPodcastView;
    private SearchResult cachedResult;


    public AddPodcastPresenter(final AddPodcastsModel addPodcastsModel, ApiLevelChecker apiLevelChecker) {
        this.addPodcastsModel = addPodcastsModel;
        this.apiLevelChecker = apiLevelChecker;
        observer = new Observer<SearchResult>() {

            @Override
            public void onCompleted() {
                showProgressBar(false);
            }

            @Override
            public void onError(Throwable throwable) {
                if (addPodcastView != null) {
                    addPodcastView.onError(throwable);
                }
                showProgressBar(false);
            }

            @Override
            public void onNext(SearchResult result) {
                if (addPodcastView != null && !result.equals(cachedResult)) {
                    cachedResult = result;
                    addPodcastView.updateSearchResults(cachedResult.getList());
                }
            }
        };
    }

    public void bindView(AddPodcastView addPodcastView, AddPodcastFragmentMemento memento) {
        this.addPodcastView = addPodcastView;
        addPodcastView.showProgressBar(memento.isProgressShowing());
        if (cachedResult != null) {
            addPodcastView.updateSearchResults(cachedResult.getList());
        }
    }

    @Override
    public void onDestroy() {
        addPodcastView = null;
    }

    @Override
    public void onStop() {
        addPodcastsModel.unsubscribeFromSearch();
    }

    @Override
    public void onStart() {
        addPodcastsModel.subscribeToSearch(observer);
    }

    public boolean onQueryTextSubmit(String query) {
        launchQuery(query);
        return true;
    }

    private void launchQuery(String query) {
        addPodcastsModel.searchPodcasts(query);
        showProgressBar(true);
    }

    public boolean onQueryTextChange(String newText) {
        addPodcastView.onQueryTextChange(newText);
        return true;
    }

    public void onSaveInstanceState(AddPodcastFragmentMemento memento) {
        memento.setProgressShowing(addPodcastView.isProgressShowing());
    }

    private void showProgressBar(boolean show) {
        if (addPodcastView != null) {
            addPodcastView.showProgressBar(show);
        }
    }

    @Override
    public void onSubscribeClicked(Podcast podcast) {
    }

    @Override
    public void onItemClicked(Podcast podcast, ImageView imageView, int position, LinearLayout titleContainer) {
        if (apiLevelChecker.isLollipopOrHigher() && !addPodcastView.isPartiallyHidden(position)) {
            addPodcastView.startDetailActivityWithTransition(podcast, imageView, titleContainer);
        } else {
            addPodcastView.startDetailActivityWithoutTransition(podcast);
        }
    }
}

File: app/src/main/java/com/pietrantuono/podcasts/addpodcast/presenter/AddPodcastPresenter.kt
package com.pietrantuono.podcasts.addpodcast.presenter

import android.arch.lifecycle.ViewModel
import android.os.Bundle
import android.widget.ImageView
import android.widget.LinearLayout
import com.pietrantuono.podcasts.GenericPresenter
import com.pietrantuono.podcasts.addpodcast.customviews.PodcastsAdapter
import com.pietrantuono.podcasts.addpodcast.model.AddPodcastsModel
import com.pietrantuono.podcasts.addpodcast.model.SearchResult
import com.pietrantuono.podcasts.addpodcast.singlepodcast.presenter.SimpleObserver
import com.pietrantuono.podcasts.addpodcast.singlepodcast.view.State
import com.pietrantuono.podcasts.addpodcast.view.AddPodcastFragment
import com.pietrantuono.podcasts.addpodcast.view.AddPodcastView
import com.pietrantuono.podcasts.addpodcast.view.ApiLevelChecker
import models.pojos.Podcast
import rx.Observer

class AddPodcastPresenter(
        private val model: AddPodcastsModel,
        private val apiLevelChecker: ApiLevelChecker)
    : ViewModel(), GenericPresenter, PodcastsAdapter.OnSunscribeClickedListener, PodcastsAdapter.OnItemClickedClickedListener {

    private val observer: Observer<SearchResult>
    private var view: AddPodcastView? = null
    private var cachedResult: SearchResult? = null
    internal val viewState: State?
        get() = model.viewState

    init {
        observer = object : SimpleObserver<SearchResult>() {
            override fun onError(throwable: Throwable) {
                onError()
            }

            override fun onNext(result: SearchResult) {
                onSearchResultsAvailable(result)
            }
        }
    }

    private fun onError() {
        setViewState(State.ERROR)
    }

    private fun onSearchResultsAvailable(result: SearchResult) {
        if (result != cachedResult) {
            cachedResult = result
            updateSearchResults(result)
        }
        setViewEmptyOrFull(result)
    }

    private fun updateSearchResults(result: SearchResult) {
        view?.updateSearchResults(result.list)
        setViewEmptyOrFull(result)
    }

    private fun setViewEmptyOrFull(result: SearchResult) {
        if (shouldBeEmpty(result)) {
            setViewState(State.EMPTY)
            return
        }
        setViewState(State.FULL)
    }

    private fun setViewState(state: State) {
        model.viewState = state
        view?.setState(state)
    }

    private fun shouldBeEmpty(result: SearchResult): Boolean = result.list == null || result.list.isEmpty()

    fun bindView(addPodcastView: AddPodcastView, bundle: Bundle?) {
        this.view = addPodcastView
        setSavedViewState(bundle)
        cachedResult?.let {
            updateSearchResults(it)
        }
    }

    private fun setSavedViewState(bundle: Bundle?) {
        val state = getViewState(bundle)
        setViewState(state)
    }

    private fun getViewState(bundle: Bundle?) = (bundle?.getSerializable(AddPodcastFragment.STATE) as? State) ?: State.EMPTY

    override fun onDestroy() {
        view = null
    }

    override fun onPause() {
        model.unsubscribeFromSearch()
    }

    override fun onResume() {
        model.subscribeToSearch(observer)
    }

    fun onQueryTextSubmit(query: String?): Boolean {
        model.searchPodcasts(query)
        return true
    }

    override fun onSubscribeClicked(podcast: Podcast) = throw UnsupportedOperationException()

    override fun onItemClicked(podcast: Podcast, imageView: ImageView, position: Int, titleContainer: LinearLayout) {
        if (apiLevelChecker.isLollipopOrHigher && view?.isPartiallyHidden(position) == false) {
            view?.startDetailActivityWithTransition(podcast, imageView, titleContainer)
        } else {
            view?.startDetailActivityWithoutTransition(podcast)
        }
    }

    fun thereIsAValidQuery() = !model.query.isNullOrBlank()

    fun refresh() {
        setViewState(State.LOADING)
        model.refresh()
    }
}
